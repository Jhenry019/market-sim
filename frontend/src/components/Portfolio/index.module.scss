@import '../../scss/root';


.portfolioInfo { margin-bottom: 1.5em; }

.return {
  margin-left: 0.5rem;
  font-size: 0.4em;
}

/** Rules for styling the portfolio return */
.flat { color: var(--gray-core); }
.up { color: var(--green-core); }
.down { color: var(--red-core); }