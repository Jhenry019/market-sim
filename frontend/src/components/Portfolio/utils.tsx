import round from "../../utils/round"


export const calculateReturn = (startAmt: number, currentAmt: number) => {
  return round((currentAmt / startAmt - 1) * 100);
}