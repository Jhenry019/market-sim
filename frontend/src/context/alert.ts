import React from 'react';


export const AlertContext = React.createContext({
  Type: "success",
  message: "",
  sendMessage: (mssg: string, Type: "success" | "error") => {}
});