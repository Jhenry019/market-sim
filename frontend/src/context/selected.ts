import React from 'react';


export const SelectedContext = React.createContext({
  symbol: '',
  nShares: 0,
  setSelected: (_: any) => {}
});