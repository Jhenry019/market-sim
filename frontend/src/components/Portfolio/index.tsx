import React, { useEffect } from 'react';
import classNames from '../../utils/classNames';
import direction from '../../utils/priceDirection';
import { calculateReturn } from './utils';
import round from '../../utils/round';
import Position, { PositionProps } from '../Position';
import styles from './index.module.scss';


export interface PortfolioProps {
  /** The amount of cash held in the portfolio */ 
  cash?: number;
  /** The total amount of equity held in the portfolio */ 
  equity?: number;
  /** The equity positions held in the portfolio */
  positions?: Array<PositionProps>;
  /** A selector used in testing environments */
  dataTestId?: string;
};

export const STARTING_CASH_AMT = 1000; 
export default function Portfolio({
  cash,
  equity,
  positions = [],
  dataTestId
}: PortfolioProps): JSX.Element 
{
  const portfolioValue = cash + equity;
  const ROI = calculateReturn(STARTING_CASH_AMT, portfolioValue);
  const DIRECTION = direction(ROI);
  const hasPositions = positions.length !== 0;

  const positionsMarkup = hasPositions && (
    <div className="pt-5">
      <h4 className="text-left">Positions</h4> 
      <ul className="list-group">
        {
          positions.map((pos: PositionProps) => (
            <Position key={pos.symbol} {...pos} />
          ))
        }
      </ul>
    </div>
  );

  const portfolioReturnMarkup = (): string => {
    if (hasPositions || ROI !== 0) return `${ROI}%`;
    return '';
  };

  const portfolioMarkup = cash !== undefined && (
    <div className={styles.portfolioInfo}>
      <h1 className="heading display-5 pb-4">
        {round(portfolioValue)}
        <span 
          className={classNames(
            styles.return,
            DIRECTION.flat && styles.flat,
            DIRECTION.up && styles.up,
            DIRECTION.down && styles.down
          )}
        >
          {portfolioReturnMarkup()}
        </span>
      </h1>
      <span>Cash: </span>
      <span>${cash}</span>
      <br />
      <span>Equity: </span>
      <span>${round(equity)}</span>
    </div>
  );

  useEffect(() => {
    // cache the portfolio data on update
    if (hasPositions || cash !== STARTING_CASH_AMT) {
      sessionStorage.setItem(
        'portfolio', 
        JSON.stringify({
          cash,
          equity, 
          positions
        })
      );
    }
  }, [portfolioValue, positions]);

  return (
    <div className="col-md-4 mx-auto" data-testid={dataTestId}>
      <div className="card card-body text-center mt-5">
        <h4 className="heading display-5 pb-4">
          My Portfolio
        </h4>
        {portfolioMarkup}
        {positionsMarkup}
      </div>
    </div>
  );
}