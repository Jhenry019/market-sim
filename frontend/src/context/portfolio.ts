import React from 'react';
import { PortfolioProps } from '../components/Portfolio';


export const PortfolioContext = React.createContext({
  cash: undefined,
  equity: undefined,
  positions: [],
  updatePortfolio: (_: PortfolioProps) => {}
});