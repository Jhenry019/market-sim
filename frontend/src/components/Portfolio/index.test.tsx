import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import { SelectedContext } from '../../context/selected';
import Portfolio from '.';


const positions = [
  {symbol: "ABC", nShares: 3, avgPrice: 11.03, dataTestId: "abc"},
  {symbol: "VOL", nShares: 2, avgPrice: 75.01, dataTestId: "vol"},
];

const context = {
  symbol: 'ABC',
  nShares: 0,
  setSelected: jest.fn()
};

it('renders a portfolio with 2 positions', () => {
  render(
    <Portfolio 
      cash={1000} 
      equity={0} 
      positions={positions}
      dataTestId="portfolio"
    />
  );
  expect(screen.getByTestId("portfolio")).toMatchSnapshot();
});

it("changes the selected instrument", () => {
  /**
   * Clicking on one of the portfolio positions
   * should change the selected instrument. 
   */
  render(
    <SelectedContext.Provider value={context}>
      <Portfolio 
        cash={1000} 
        equity={0} 
        positions={positions}
        dataTestId="portfolio"
      />
    </SelectedContext.Provider>
  );
  
  fireEvent.click(screen.getByTestId("vol"));
  expect(context.setSelected).toBeCalledWith({
    ...context,
    symbol: "VOL",
    nShares: 2
  });
});