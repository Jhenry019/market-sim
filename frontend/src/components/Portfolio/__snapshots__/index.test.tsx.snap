// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a portfolio with 2 positions 1`] = `
<div
  class="col-md-4 mx-auto"
  data-testid="portfolio"
>
  <div
    class="card card-body text-center mt-5"
  >
    <h4
      class="heading display-5 pb-4"
    >
      My Portfolio
    </h4>
    <div
      class="portfolioInfo"
    >
      <h1
        class="heading display-5 pb-4"
      >
        1000
        <span
          class="return flat"
        >
          0%
        </span>
      </h1>
      <span>
        Cash: 
      </span>
      <span>
        $
        1000
      </span>
      <br />
      <span>
        Equity: 
      </span>
      <span>
        $
        0
      </span>
    </div>
    <div
      class="pt-5"
    >
      <h4
        class="text-left"
      >
        Positions
      </h4>
      <ul
        class="list-group"
      >
        <li
          class="ticker"
          data-testid="abc"
        >
          <div
            class="ml-3"
          >
            <span
              class="row"
            >
              ABC
            </span>
            <span
              class="row"
            >
              3
               shares @ $
              11.03
              /share
            </span>
          </div>
          <span
            class="pill down"
          >
            0
          </span>
        </li>
        <li
          class="ticker"
          data-testid="vol"
        >
          <div
            class="ml-3"
          >
            <span
              class="row"
            >
              VOL
            </span>
            <span
              class="row"
            >
              2
               shares @ $
              75.01
              /share
            </span>
          </div>
          <span
            class="pill down"
          >
            0
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
